sql/src/main/java/io/crate/planner/operators/ExplainLogicalPlan.java

```diff
@@ -0,0 +1,199 @@
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.planner.operators;
+
+import com.google.common.collect.ImmutableMap;
+import io.crate.analyze.OrderBy;
+import io.crate.analyze.symbol.Function;
+import io.crate.analyze.symbol.SelectSymbol;
+import io.crate.analyze.symbol.Symbol;
+import io.crate.analyze.symbol.format.SymbolPrinter;
+import io.crate.execution.dsl.projection.builder.ProjectionBuilder;
+import io.crate.planner.PlannerContext;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static io.crate.planner.PlanPrinter.refs;
+
+public class ExplainLogicalPlan {
+
+    private static final Visitor VISITOR = new Visitor();
+
+    public static Map<String, Object> objectMap(LogicalPlan logicalPlan,
+                                                PlannerContext plannerContext,
+                                                ProjectionBuilder projectionBuilder) {
+        return VISITOR.process(logicalPlan, new Context(plannerContext, projectionBuilder)).build();
+    }
+
+    private static class Context {
+        private final PlannerContext plannerContext;
+        private final ProjectionBuilder projectionBuilder;
+
+        public Context(PlannerContext plannerContext, ProjectionBuilder projectionBuilder) {
+            this.plannerContext = plannerContext;
+            this.projectionBuilder = projectionBuilder;
+        }
+    }
+
+    private static class Visitor extends LogicalPlanVisitor<Context, ImmutableMap.Builder<String, Object>> {
+
+        private static ImmutableMap.Builder<String, Object> newBuilder() {
+            return ImmutableMap.builder();
+        }
+
+        private static ImmutableMap.Builder<String, Object> subMap() {
+            return newBuilder().put("type", "logicalOperator");
+        }
+
+        private static ImmutableMap.Builder<String, Object> toMap(LogicalPlan logicalPlan,
+                                                                  ImmutableMap.Builder<String, Object> subMap) {
+            return newBuilder()
+                .put(logicalPlan.getClass().getSimpleName(), subMap.build());
+        }
+
+        private static Map<String, Object> explainMap(LogicalPlan logicalPlan, Context context) {
+            return logicalPlan.explainMap(context.plannerContext, context.projectionBuilder);
```

Comment:
I think it's a bit confusing that we're mixing 2 different dispatching patterns here. On one hand there is the externalized visitor dispatching. On the other hand there is the `explainMap` on the LogicalPlan itself. 

Couldn't we get rid of one of those and just use one pattern, e.g. by putting the logic of `LogicalPlan.explainMap` also into the visitor?